common/template_string.h

```diff
@@ -0,0 +1,80 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_TEMPLATE_STRING_H_
+#define CARBON_COMMON_TEMPLATE_STRING_H_
+
+#include "llvm/ADT/StringRef.h"
+
+namespace Carbon {
+
+// Represent a compile-time string in a form suitable for non-type template
+// arguments.
+//
+// These arguments are required to be a "structural type", and so we copy the
+// string contents into a public array of `char`s. For details, see:
+// https://en.cppreference.com/w/cpp/language/template_parameters#Non-type_template_parameter
+//
+// Designed to support implicitly deduced construction from a string literal
+// template argument. This type will implicitly convert to an `llvm::StringRef`
+// for accessing the string contents, and also provides a dedicated `c_str()`
+// method to access the string as a C-string.
+//
+// Example usage:
+// ```cpp
+// template <TemplateString Str> auto F() -> void {
+//   llvm::cout() << Str;
+// }
+//
+// auto Example() -> void {
+//   F<"string contents here">();
+// }
+// ```
+template <int N>
+struct TemplateString {
+  // Constructs the template string from a string literal.
+  //
+  // Intentionally implicit conversion from string literals for use as a
+  // non-type template parameter.
+  //
+  // The closest we can get to explicitly accepting a string literal is to
+  // accept an array of `const char`s, so we additionally use Clang's constexpr
+  // `enable_if` attribute to require the array to be usable as a C-string with
+  // the expected length. This checks both for null-termination and no embedded
+  // `0` bytes.
+  //
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  constexpr TemplateString(const char (&str)[N + 1]) __attribute__((
+      enable_if(__builtin_strlen(str) == N,
+                "character array is not nul-terminated valid C-string"))) {
```

Comment:
"nul" here, "null" above. (I guess this one was missed when addressing Josh's prior comment.)
```suggestion
                "character array is not null-terminated valid C string"))) {
```